File: GenerateParentheses.py
"""
22. Generate Parentheses
Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.
For example, given n = 3, a solution set is:

"((()))", "(()())", "(())()", "()(())", "()()()"
"""
#Catalan number
# O(4^n / n^(3/2))
# O(n)
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        if n <= 0:
            return []
        res = []
        self.recur(n, "", 0, 0, res)
        return res
        
    def recur(self, n, cur, left, right, res):
        if left + right == 2 * n:
    #in this recurtion, no duplicate will be generated
            res.append(cur[:])
            return 
        if left > right:
            self.recur(n, cur + ")", left, right + 1, res)
        if left < n:
            self.recur(n, cur + "(", left + 1, right, res)

if __name__ == "__main__":
    print(len(Solution().generateParenthesis(7)))

File: Java/GenerateParentheses.java
import java.util.*;

public class GenerateParentheses {

    public static List<String> generateParenthesis(int n) {
        int left = 0;
        int right = 0;
        StringBuffer cur = new StringBuffer();
        List<String> res = new ArrayList<String>();
        if(n != 0) {
            recur(res, cur, left, right, n);
        }
        return res;
    }
    
    public static void recur(List<String> res, StringBuffer cur, int left, int right, int n) {
        if(left == right && left == n) {
            res.add(cur.toString());
            return;
        }
        if(left > right) {
            cur.append(")");
            recur(res, cur, left, right + 1, n);
            cur.deleteCharAt(cur.length() - 1);
        } 
        if(left < n) {
            cur.append("(");
            recur(res, cur, left + 1, right, n);
            cur.deleteCharAt(cur.length() - 1);
        }
    }
    
	public static void main(String[] args) {
		System.out.println(generateParenthesis(4));

	}

}
